### PR TITLE
refactor: handle apiKey and subscriptions caches using cacheManagers instead of repositoryWrappers

### PR DESCRIPTION
**Pull request by marcambier**
**Issue**

https://github.com/gravitee-io/issues/issues/7360

**Description**

feat: handle apiKey and subscriptions with dedicated services instead of repositoryWrappers

Policies should never access to underlying repository directly,
So they will use those new services instead of repositoryWrappers.

This introduces a new CheckSubscriptionPolicy for API keys authentication, dedicated to check the subscription dates validity.
Like it's already implemented for JWT and Oauth security types.

Subscription validity check is also implemented in Jupiter SecurityPlanHandler, via the new service.

If associated subscription is not yet started or already expired, request will be unauthorized.

This permits to remove dirty workarounds in SubscriptionService : we don't need nomore to force api key cache refreshment,
Cause subscription will be checked from dedicated subscription cache.

🚀 CI was able to deploy the build of this PR, so you can now try it directly [here](https://apimnightlywebui24386.z6.web.core.windows.net/refactor-apikeysandsubsmanagers/index.html)
_Notes_: The deployed app is linked to the management API of APIM master. (Same login and password as APIM master)

---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-nbmeckxced.chromatic.com)

### CODE REVIEW

File: gravitee-apim-gateway/gravitee-apim-gateway-security/gravitee-apim-gateway-security-apikey/src/test/java/io/gravitee/gateway/security/apikey/policy/CheckSubscriptionPolicyTest.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        when(executionContext.getComponent(SubscriptionCacheManager.class)).thenReturn(subscriptionCacheManager);
+    }
+
+    @Test
+    public void should_fail_unauthorized_cause_no_subscription_found_in_cache() throws PolicyException {
+        when(subscriptionCacheManager.getById(SUBSCRIPTION_ID)).thenReturn(Optional.empty());
+
+        policy.execute(policyChain, executionContext);
+
+        verify(policyChain, times(1))
```
Comment (Okhelifi) [inline gravitee-apim-gateway/gravitee-apim-gateway-security/gravitee-apim-gateway-security-apikey/src/test/java/io/gravitee/gateway/security/apikey/policy/CheckSubscriptionPolicyTest.java]: For the verify ( all in this test file ) the default time is 1 so you can remove it if you want.

File: gravitee-apim-gateway/gravitee-apim-gateway-handlers/gravitee-apim-gateway-handlers-api/src/main/java/io/gravitee/gateway/handlers/api/cache/ApiKeyCacheManager.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    public Optional<ApiKey> get(String api, String key) {
+        return Optional.ofNullable(cache.get(buildCacheKey(api, key)));
+    }
+
+    private String buildCacheKey(ApiKey apiKey) {
+        return buildCacheKey(apiKey.getApi(), apiKey.getKey());
+    }
+
+    private String buildCacheKey(String api, String key) {
+        return String.format("%s.%s", api, key);
```
Comment (Okhelifi) [inline gravitee-apim-gateway/gravitee-apim-gateway-handlers/gravitee-apim-gateway-handlers-api/src/main/java/io/gravitee/gateway/handlers/api/cache/ApiKeyCacheManager.java]: I would have align the cache keys constructions with the subscriptions

```suggestion
        return api + '-' + key;
```

File: gravitee-apim-gateway/gravitee-apim-gateway-security/gravitee-apim-gateway-security-oauth2/src/main/java/io/gravitee/gateway/security/oauth2/policy/CheckSubscriptionPolicy.java [hunk 3]
```diff
@@ -45,18 +41,20 @@ public class CheckSubscriptionPolicy implements Policy {
...
     @Override
     public void execute(PolicyChain policyChain, ExecutionContext executionContext) throws PolicyException {
-        SubscriptionRepository subscriptionRepository = executionContext.getComponent(SubscriptionRepository.class);
-
-        // Get plan and client_id from execution context
+        // get stuff from execution context
+        SubscriptionCacheManager subscriptionCacheManager = executionContext.getComponent(SubscriptionCacheManager.class);
+        String api = (String) executionContext.getAttribute(ExecutionContext.ATTR_API);
         String clientId = (String) executionContext.getAttribute(CONTEXT_ATTRIBUTE_CLIENT_ID);
+        final String plan = (String) executionContext.getAttribute(ExecutionContext.ATTR_PLAN);
```
Comment (Okhelifi) [inline gravitee-apim-gateway/gravitee-apim-gateway-security/gravitee-apim-gateway-security-oauth2/src/main/java/io/gravitee/gateway/security/oauth2/policy/CheckSubscriptionPolicy.java]:
```suggestion
       String plan = (String) executionContext.getAttribute(ExecutionContext.ATTR_PLAN);
```

**Review by Okhelifi — COMMENTED** [review]
Like discussed I would do some test with you and then I would approve if you are ok 😅

File: gravitee-apim-gateway/gravitee-apim-gateway-handlers/gravitee-apim-gateway-handlers-api/src/main/java/io/gravitee/gateway/handlers/api/cache/ApiKeyCacheManager.java (hunk 2 above)
Comment (marcambier) [reply gravitee-apim-gateway/gravitee-apim-gateway-handlers/gravitee-apim-gateway-handlers-api/src/main/java/io/gravitee/gateway/handlers/api/cache/ApiKeyCacheManager.java]: Nice catch.
I did the contrary : construct the subscription key the same way as api keys, with the ".";
I prefer the "." as separator cause uuid already contains "-", it's visually simpler to read.

**Review by Okhelifi — APPROVED** [review]

**Review by Okhelifi — APPROVED** [review]
Tested with you. Still lgtm Good job 👍🏼

**Pull request merged**